[Merged by Bors] - feat(CategoryTheory/Closed/Enrichment): a closed monoidal category is enriched in itself

daniel-carranza opened the pull request:
From a monoidal closed category V, construct an instance of `EnrichedCategory V` on V where the hom-objects are given by the internal homs of the closed structure.

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

First time contributor. I am still new to formalization, so I am open any and all suggestions! In particular, the proofs here are a bit long compared to most proofs I see in the library, and I'm not sure if it's couth to use the `@[simp]` attribute for some equalities here.

Remark: As I understand, the type of simplicial sets `SSet.{v}` also has an instance of `EnrichedCategory SSet.{v}` constructed in `AlgebraicTopology/SimplicialCategory/SimplicialObject`. This enrichment comes from viewing `SSet.{v}` as a category of simplicial objects, rather than as a closed monoidal category. I'm still a beginner, but could this cause any issues if a file imports both instances?

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

b-mehta reviewed (commented):
This is in pretty good shape already for a first contribution! I've left a couple of style comments.
Note that the title of your PR shouldn't be capitalised, see https://leanprover-community.github.io/contribute/commit.html

emilyriehl commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,119 @@
+def id (x : V) : 𝟙_ V ⟶ (ihom x).obj x := curry (ρ_ x).hom
+
+/-- The *uncurried* composition morphism
+  x ⊗ (hom(x, y) ⊗ hom(y, z)) ⟶ (x ⊗ hom(x, y)) ⊗ hom(y, z) ⟶ y ⊗ hom(y, z) ⟶ z.
```
I think these render slightly better if you use single backticks ``lean code`` around the verbatim lean code.

emilyriehl reviewed (commented):
@b-mehta is a way more expert reviewer than me. I'm curious about the answer to the dual instance question you raised. Maybe ask it on the zulip of @mckoen who provided the other instance (iirc)?

b-mehta commented:
I think the dual instance could cause issues, but I don't know enough about the context of this to know what the correct solution is. I'd guess we'd like to only have one of them actually active on SSet, prove that they're the same or make the constructions so they're definitionally the same. But I don't know the maths here well enough to be sure (are they even the same, informally?). Asking on the zulip is definitely the correct thing to do.

joelriou commented:
About the dual instance issue, I would suggest making the new instance in this PR a local or a scoped instance.

daniel-carranza commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,119 @@
+  rw [uncurry_curry, uncurry_curry]; dsimp
+  rw [associator_inv_naturality_middle_assoc]
+  rw [← comp_whiskerRight_assoc]
+  rw [← uncurry_eq, uncurry_curry]
```
Indeed, this pair of rewrites is exactly that equality. The current proof is the best I could come up with in terms of cleaning on my end, but my hope is also that there should be a cleaner proof!

daniel-carranza replied:
I experimented (locally on my machine) with making this a separate lemma inside the `CategoryTheory/Closed/Monoidal.lean` file with the `simp` attribute. The proof now goes through using a single simp wherever these two rewrites originally occurred. Is this a good candidate simp lemma to be added to the CategoryTheory/Closed/Monoidal file? (Posted on Zulip as well)

daniel-carranza commented:
I've implemented Joel's suggestion in the current version, but I created a Zulip chat to discuss this in the <a href="https://leanprover.zulipchat.com/#narrow/stream/455414-Infinity-Cosmos/topic/Closed.20Monoidal.20cats.20are.20self-enriched"> Infinity-Cosmos </a> channel

daniel-carranza commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,119 @@
+  rw [← uncurry_eq, uncurry_curry];
+  simp only [whiskerLeft_rightUnitor_inv, MonoidalCategory.whiskerRight_id, Category.assoc,
+    Iso.inv_hom_id, Category.comp_id, Iso.hom_inv_id_assoc, Iso.inv_hom_id_assoc]
+  exact Eq.symm (uncurry_id_eq_ev x y)
```
done

daniel-carranza commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,119 @@
+lemma comp_eq (x y z : V) : comp x y z = curry (compTranspose x y z) := rfl
+
+/-- Associativity of V-composition. Used to prove that V is enriched in itself. -/
+theorem assoc (x y z w : V) : (α_ _ _ _).inv ≫ comp x y z ▷ _ ≫ comp x z w =
```
done

daniel-carranza commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,119 @@
+  rw [← uncurry_eq, uncurry_curry]
+  rw [associator_inv_naturality_right_assoc]
+  rw [whisker_exchange_assoc]
+  rw [← uncurry_eq, uncurry_curry]
```
done

daniel-carranza commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,119 @@
+lemma compTranpose_eq (x y z : V) : compTranspose x y z = (α_ _ _ _).inv ≫
+    (ihom.ev x).app y ▷ _ ≫ (ihom.ev y).app z :=
+  rfl
+
```
done

daniel-carranza commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,119 @@
+
+/-- Unfold the definition of id.
+This exists to streamline the proofs of MonoidalClosed.id_comp and MonoidalClosed.comp_id -/
+@[simp]
```
done

daniel-carranza commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,119 @@
+theorem id_comp (x y : V) : (λ_ _).inv ≫ id x ▷ (ihom x).obj y ≫ comp x x y = 𝟙 _ := by
+  apply uncurry_injective; simp only [uncurry_natural_left, comp_eq, id_eq]
+  rw [uncurry_curry]
+  rw [whisker_assoc_symm];
```
done

daniel-carranza commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,119 @@
+
+/-- Left unitality of V-composition. Used to prove that V is enriched in itself. -/
+theorem id_comp (x y : V) : (λ_ _).inv ≫ id x ▷ (ihom x).obj y ≫ comp x x y = 𝟙 _ := by
+  apply uncurry_injective; simp only [uncurry_natural_left, comp_eq, id_eq]
```
done

daniel-carranza commented:
@b-mehta  @joelriou Sorry for the ping! I'm wondering if anyone could take another look at this PR for suggestions/how to move towards merging the changes - thank you in advance!

b-mehta commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,99 @@
+/-- The V-composition morphism
+  `hom(x, y) ⊗ hom(y, z) ⟶ hom(x, z)`
+used to equip V with the structure of a V-category -/
+def comp (x y z : V) := curry (compTranspose x y z)
```
Let's put a type ascription on this for readability.
```suggestion
def comp (x y z : V) : (ihom x).obj y ⊗ (ihom y).obj z ⟶ (ihom x).obj z := 
  curry (compTranspose x y z)
```

b-mehta commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,99 @@
+
+/-- Unfold the definition of compTranspose.
+This exists to streamline the proof of MonoidalClosed.assoc -/
+lemma compTranpose_eq (x y z : V) :
```
```suggestion
lemma compTranspose_eq (x y z : V) :
```

b-mehta commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,99 @@
+/-- The *uncurried* composition morphism
+  `x ⊗ (hom(x, y) ⊗ hom(y, z)) ⟶ (x ⊗ hom(x, y)) ⊗ hom(y, z) ⟶ y ⊗ hom(y, z) ⟶ z`.
+The V-composition morphism is defined as the adjoint transpose of this map. -/
+def compTranspose (x y z : V) :=
+  (α_ x ((ihom x).obj y) ((ihom y).obj z)).inv ≫
+    (ihom.ev x).app y ▷ ((ihom y).obj z) ≫
+    (ihom.ev y).app z
```
```suggestion
def compTranspose (x y z : V) : x ⊗ (ihom x).obj y ⊗ (ihom y).obj z ⟶ z :=
  (α_ x ((ihom x).obj y) ((ihom y).obj z)).inv ≫
    (ihom.ev x).app y ▷ ((ihom y).obj z) ≫
    (ihom.ev y).app z
```

b-mehta commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,99 @@
+
+/-- The *uncurried* composition morphism
+  `x ⊗ (hom(x, y) ⊗ hom(y, z)) ⟶ (x ⊗ hom(x, y)) ⊗ hom(y, z) ⟶ y ⊗ hom(y, z) ⟶ z`.
+The V-composition morphism is defined as the adjoint transpose of this map. -/
```
```suggestion
The V-composition morphism will be defined as the adjoint transpose of this map. -/
```
if you like

b-mehta commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,99 @@
+  id := id
+  comp := comp
+  id_comp := fun _ _ => by
+    apply uncurry_injective; simp only [uncurry_natural_left, comp_eq, id_eq]
```
```suggestion
    apply uncurry_injective
    simp only [uncurry_natural_left, comp_eq, id_eq]
```
It's almost always preferred to have one tactic per line, like this

b-mehta commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,99 @@
+  id := id
+  comp := comp
+  id_comp := fun _ _ => by
+    apply uncurry_injective; simp only [uncurry_natural_left, comp_eq, id_eq]
+    rw [uncurry_curry, whisker_assoc_symm]; simp only [compTranpose_eq, Category.assoc]
+    rw [Iso.hom_inv_id_assoc, ← comp_whiskerRight_assoc, ← uncurry_eq, uncurry_curry]
+    simp only [Functor.id_obj, triangle_assoc_comp_right_assoc, whiskerLeft_inv_hom_assoc]
+    exact (uncurry_id_eq_ev _ _).symm
```
```suggestion
    apply uncurry_injective
    rw [uncurry_natural_left, uncurry_natural_left, comp_eq, uncurry_curry, id_eq, compTranspose_eq,
      associator_inv_naturality_middle_assoc, ← comp_whiskerRight_assoc, ← uncurry_eq,
      uncurry_curry, triangle_assoc_comp_right_assoc, whiskerLeft_inv_hom_assoc,
      uncurry_id_eq_ev _ _]
```
Here's a slightly more condensed proof - entirely optional if you take it! The only real difference is in the use of `associator_inv_naturality_middle_assoc`, which lets me avoid dealing with cancelling out isomorphisms

b-mehta commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,99 @@
+
+/-- For V closed monoidal, build an instance of V as a V-category -/
+scoped instance : EnrichedCategory V V where
+  Hom := fun x => (ihom x).obj
```
```suggestion
  Hom x := (ihom x).obj
```

b-mehta commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,99 @@
+  Hom := fun x => (ihom x).obj
+  id := id
+  comp := comp
+  id_comp := fun _ _ => by
```
```suggestion
  id_comp _ _ := by
```
these mean the same thing, but the shortcut is a bit nicer :)

b-mehta commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,99 @@
+    simp only [Functor.id_obj, triangle_assoc_comp_right_assoc, whiskerLeft_inv_hom_assoc]
+    exact (uncurry_id_eq_ev _ _).symm
+  comp_id := fun _ _ => by
+    apply uncurry_injective
+    simp only [uncurry_natural_left, comp_eq, id_eq]
+    rw [uncurry_curry]; simp only [compTranpose_eq, Category.assoc]
+    rw [associator_inv_naturality_right_assoc, whisker_exchange_assoc]; dsimp
+    rw [← uncurry_eq, uncurry_curry]
+    simp only [whiskerLeft_rightUnitor_inv, MonoidalCategory.whiskerRight_id, Category.assoc,
+      Iso.inv_hom_id, Category.comp_id, Iso.hom_inv_id_assoc, Iso.inv_hom_id_assoc]
+    exact (uncurry_id_eq_ev _ _).symm
```
```suggestion
    apply uncurry_injective
    rw [uncurry_natural_left, uncurry_natural_left, comp_eq, uncurry_curry, compTranspose_eq,
      associator_inv_naturality_right_assoc, ← rightUnitor_tensor_inv_assoc,
      whisker_exchange_assoc, ← rightUnitor_inv_naturality_assoc, ← uncurry_id_eq_ev y y,
      ← uncurry_natural_left]
    simp [id_eq, uncurry_id_eq_ev]
```
also optional

b-mehta reviewed (approved):
Sorry for the delay on this. It's in very good shape overall, I made a few readability suggestions and optional alternate proofs.

joelriou commented:
Most of the definitions and compatibilities in this PR could be phrased under the slightly weaker condition that only a few `Closed X` assumptions hold. Then, I believe that the definitions and lemmas should be moved to the file `Closed.Monoidal` along with the other compatibilities which were already obtained. Then, `Closed.Enrichment` should become a very short file which applies this API when `MonoidalClosed` holds.

daniel-carranza commented:
Thanks everyone for the suggestions! I've implemented the simplifications and moved all the lemmas I could to the `Closed/Monoidal` file. I've also updated the comments/docstrings to reflect the new file structure

joelriou commented on Mathlib/CategoryTheory/Closed/Monoidal.lean:
```diff
@@ -304,6 +304,87 @@ theorem ofEquiv_uncurry_def {X Y Z : C} :
+The proofs of associativity and unitality use the following outline:
+  1. Take adjoint transpose on each side of the equality (uncurry_injective)
+  2. Do whatever rewrites/simps are necessary to apply uncurry_curry
+  3. Conclude with simp (+ a possible exact)
```
Turns out no exact is necessary after simp:
```suggestion
  3. Conclude with simp
```

joelriou commented on Mathlib/CategoryTheory/Closed/Monoidal.lean:
```diff
@@ -304,6 +304,87 @@ theorem ofEquiv_uncurry_def {X Y Z : C} :
+-/
+
+/-- Left unitality of the enriched structure -/
+lemma id_comp (x y : C) [Closed x] :
```
```suggestion
@[reassoc (attr := simp)]
lemma id_comp (x y : C) [Closed x] :
```

joelriou commented on Mathlib/CategoryTheory/Closed/Monoidal.lean:
```diff
@@ -304,6 +304,87 @@ theorem ofEquiv_uncurry_def {X Y Z : C} :
+      uncurry_id_eq_ev _ _]
+
+/-- Right unitality of the enriched structure -/
+lemma comp_id (x y : C) [Closed x] [Closed y] :
```
```suggestion
@[reassoc (attr := simp)]
lemma comp_id (x y : C) [Closed x] [Closed y] :
```

joelriou commented on Mathlib/CategoryTheory/Closed/Monoidal.lean:
```diff
@@ -304,6 +304,87 @@ theorem ofEquiv_uncurry_def {X Y Z : C} :
+  simp [id_eq, uncurry_id_eq_ev]
+
+/-- Associativity of the enriched structure -/
+lemma assoc (w x y z : C) [Closed w] [Closed x] [Closed y] :
```
```suggestion
@[reassoc]
lemma assoc (w x y z : C) [Closed w] [Closed x] [Closed y] :
```

joelriou commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,45 @@
+/-- For C closed monoidal, build an instance of C as a C-category -/
+scoped instance : EnrichedCategory C C where
+  Hom x := (ihom x).obj
+  id x := @id C _ _ x _
+  comp x y z := @comp C _ _ x y z _ _
+  id_comp x y := @id_comp C _ _ x y _
+  comp_id x y := @comp_id C _ _ x y _ _
+  assoc x y z w := @assoc C _ _ x y z w _ _ _
```
```suggestion
  id _ := id _
  comp _ _ _ := comp _ _ _
  assoc _ _ _ _ := assoc _ _ _ _
```

joelriou commented on Mathlib/CategoryTheory/Closed/Monoidal.lean:
```diff
@@ -304,6 +304,87 @@ theorem ofEquiv_uncurry_def {X Y Z : C} :
+The C-composition morphism will be defined as the adjoint transpose of this map. -/
+def compTranspose (x y z : C) [Closed x] [Closed y] : x ⊗ (ihom x).obj y ⊗ (ihom y).obj z ⟶ z :=
+  (α_ x ((ihom x).obj y) ((ihom y).obj z)).inv ≫
+    (ihom.ev x).app y ▷ ((ihom y).obj z) ≫
+    (ihom.ev y).app z
```
```suggestion
    (ihom.ev x).app y ▷ ((ihom y).obj z) ≫ (ihom.ev y).app z
```

joelriou commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,43 @@
+
+variable {C : Type u} [Category.{v} C] [MonoidalCategory C] [MonoidalClosed C]
+
+/-- For C closed monoidal, build an instance of C as a C-category -/
```
```suggestion
/-- For `C` closed monoidal, build an instance of `C` as a `C`-category -/
```

joelriou commented on Mathlib/CategoryTheory/Closed/Monoidal.lean:
```diff
@@ -304,6 +304,89 @@ theorem ofEquiv_uncurry_def {X Y Z : C} :
+
+/-- The C-identity morphism
+  `𝟙_ C ⟶ hom(x, x)`
+used to equip C with the structure of a C-category -/
```
```suggestion
used to equip `C` with the structure of a `C`-category -/
```

joelriou commented on Mathlib/CategoryTheory/Closed/Monoidal.lean:
```diff
@@ -304,6 +304,89 @@ theorem ofEquiv_uncurry_def {X Y Z : C} :
+
+/-- The *uncurried* composition morphism
+  `x ⊗ (hom(x, y) ⊗ hom(y, z)) ⟶ (x ⊗ hom(x, y)) ⊗ hom(y, z) ⟶ y ⊗ hom(y, z) ⟶ z`.
+The C-composition morphism will be defined as the adjoint transpose of this map. -/
```
```suggestion
The `C`-composition morphism will be defined as the adjoint transpose of this map. -/
```

joelriou commented on Mathlib/CategoryTheory/Closed/Monoidal.lean:
```diff
@@ -304,6 +304,89 @@ theorem ofEquiv_uncurry_def {X Y Z : C} :
+  (α_ x ((ihom x).obj y) ((ihom y).obj z)).inv ≫
+    (ihom.ev x).app y ▷ ((ihom y).obj z) ≫ (ihom.ev y).app z
+
+/-- The C-composition morphism
+  `hom(x, y) ⊗ hom(y, z) ⟶ hom(x, z)`
+used to equip C with the structure of a C-category -/
```
```suggestion
/-- The `C`-composition morphism
  `hom(x, y) ⊗ hom(y, z) ⟶ hom(x, z)`
used to equip `C` with the structure of a `C`-category -/
```

joelriou commented:
After the kind of style suggestions, I think this PR is ready.

daniel-carranza commented:
Thank you!

joelriou commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,43 @@
+`SSet.{v}` has an instance of `EnrichedCategory SSet.{v}` as a category of simplicial objects;
+see `AlgebraicTopology/SimplicialCategory/SimplicialObject`).
+
+All structure field values are defined in `Closed/Monoidal`
```
```suggestion
All structure field values are defined in `Closed/Monoidal`.
```

joelriou commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,43 @@
+
+The instance is given at the end of the file.
+
+We use `scoped instance` to avoid potential issues where C may also have
+a C-category structure coming from another source (e.g. the type of simplicial sets
+`SSet.{v}` has an instance of `EnrichedCategory SSet.{v}` as a category of simplicial objects;
+see `AlgebraicTopology/SimplicialCategory/SimplicialObject`).
```
```suggestion
We use `scoped instance` to avoid potential issues where `C` may also have
a `C`-category structure coming from another source (e.g. the type of simplicial sets
`SSet.{v}` has an instance of `EnrichedCategory SSet.{v}` as a category of simplicial objects;
see `AlgebraicTopology/SimplicialCategory/SimplicialObject`).
```

joelriou commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,43 @@
+From the data of a closed monoidal category C, we define a C-category structure for C
+where the hom-object is given by the internal hom (coming from the closed structure).
+
+The instance is given at the end of the file.
+
```
```suggestion
```

joelriou commented on Mathlib/CategoryTheory/Closed/Enrichment.lean:
```diff
@@ -0,0 +1,43 @@
+/-!
+# A closed monoidal category is enriched in itself
+
+From the data of a closed monoidal category C, we define a C-category structure for C
```
```suggestion
From the data of a closed monoidal category `C`, we define a `C`-category structure for `C`.
```

joelriou commented:
Thanks!

bors merge
